### PR TITLE
DL for Key Authority Declaration

### PR DESCRIPTION
**Pull request by mjbrisebois**
Exploring the options for the `key_authority` declaration in app chains that is required to fix issue #23

### CODE REVIEW

**Comment by mjbrisebois** [comment]
I like the `ActionHash` option pointing to the KSR better as it allows the check to just confirm that the creator of a key anchor was a member of the authority.  However, I haven't thought of a solution for guaranteeing that a member followed the latest change rules.  Still thinking that through...

File: docs/DL-2024-05-20-key-authority-declaration.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+
+### 2. `key_authority: ActionHash`
+
+Points to the KSR authority.
+
+#### Validation process when claiming an app agent
+
+- Check that the `KeyAnchor` in deepkey has a create made by an agent that is a member of the
+  `key_authority` (KSR)
```
Comment (maackle) [inline docs/DL-2024-05-20-key-authority-declaration.md:50]: Is the idea here that we could be checking for authorship by multiple agents, hence allowing new registrations by multiple agents? Is there an advantage to this?

**Review by maackle — COMMENTED** [review]
I'm not clear on what the advantage would be of one of the other. The first option seems simpler and there are less open questions, and also solves the circumvention issue, so what is the benefit of the second option?

File: docs/DL-2024-05-20-key-authority-declaration.md (hunk 1 above)
Comment (mjbrisebois) [reply docs/DL-2024-05-20-key-authority-declaration.md:50]: Yes, this follows the original design of deepkey to have multiple devices all managing the same keyset.  The only part about this solution that seems problematic to me is how it changes the Holochain integration spec from being a generic `AgentPubKey` comparison to specific app knowledge about the KSR and invite acceptance entry structs.  It is possible to lookup the KSR using the `AgentPubKey` which means option 1 could evolve to option 2 without changing the `key_authority` hash type.

File: docs/DL-2024-05-20-key-authority-declaration.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  - Eg. for each create, check that their KSR, or their latest invite acceptance matches the KSR
+    authority.
+
+#### Validation process when updating an app agent
+
+- Lookup the previous agent record to get the `key_authority`
+- Check that the `KeyAnchor` in deepkey has a create made by an agent that is a member of the
+  `key_authority` (KSR)
+  - Eg. for each create, check that their KSR, or their latest invite acceptance matches the KSR
```
Comment (artbrock) [inline docs/DL-2024-05-20-key-authority-declaration.md:59]: I thought this is one of the main issues about why this approach doesn't work. Validation code can't check for "latest" invite acceptance, only for elements in the chain preceding the item being validated. 

Is that what you mean by latest? Most recent preceding the item being validated? (Also in line 51 above.)

Comment (mjbrisebois) [reply docs/DL-2024-05-20-key-authority-declaration.md:59]: Yes, it would be on the same chain so validation can look backward from the key creation.

Comment (mjbrisebois) [reply docs/DL-2024-05-20-key-authority-declaration.md:59]:
```suggestion
  - Eg. for each create, check that their KSR, or the latest invite acceptance on their chain, matches the KSR
```

File: docs/DL-2024-05-20-key-authority-declaration.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+### 2. `key_authority: ActionHash`
+
+Points to the KSR authority.
+
+#### Validation process when claiming an app agent
+
+- Check that the `KeyAnchor` in deepkey has a create made by an agent that is a member of the
+  `key_authority` (KSR)
+  - Eg. for each create, check that their KSR, or their latest invite acceptance matches the KSR
```
Comment (mjbrisebois) [inline docs/DL-2024-05-20-key-authority-declaration.md:51]:
```suggestion
  - Eg. for each create, check that their KSR, or the latest invite acceptance on their chain, matches the KSR
```

**Comment by mjbrisebois** [comment]
Going with option 1 `key_authority: AgentPubKey`

**Pull request closed without merging**